### PR TITLE
Splicing gather cleanup

### PR DESCRIPTION
**Pull request by goedderz**
### Scope & Purpose

Cleanup of things found during #10341.

Amongst other things, this adds `-Wno-error=deprecated-declarations` and `-Wno-deprecated-declarations`, so we may use `[[deprecated]]` without compilation failures or a spammed log. It can be enabled locally by passing `-DCMAKE_CXX_FLAGS="-Wdeprecated-declarations"`.

It also enables the `splice-subqueries` rule when ModificationBlocks are involved.

### CODE REVIEW

**Comment by goedderz** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/7277/

**Comment by goedderz** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/7290/

**Comment by goedderz** [comment]
Jenkins is blue.

File: arangod/Aql/ExecutionBlockImpl.cpp [hunk 1]
```diff
@@ -49,6 +49,7 @@
 #include "Aql/ModificationExecutor.h"
 #include "Aql/MultiDependencySingleRowFetcher.h"
 #include "Aql/NoResultsExecutor.h"
+#include "Aql/ParallelUnsortedGatherExecutor.h"
```
Comment (markuspf) [inline arangod/Aql/ExecutionBlockImpl.cpp]: This looks like a rather odd change here; does this belong to another PR?

File: arangod/Aql/IdExecutor.h [hunk 2]
```diff
@@ -31,7 +31,20 @@
 #include <tuple>
 #include <utility>
 
-// TODO Clean up unused variants of the IdExecutor - some of them aren't in use anymore.
+// There are currently three variants of IdExecutor in use:
```
Comment (markuspf) [inline arangod/Aql/IdExecutor.h]: Yay!

File: arangod/Aql/OptimizerRules.cpp [hunk 3]
```diff
@@ -7283,7 +7283,7 @@ namespace {
 
 bool nodeMakesThisQueryLevelUnsuitableForSubquerySplicing(ExecutionNode const* const node) {
   // TODO Enable modification nodes again, as soon as the corresponding branch
-  //  is merged.
+  //  is merged. Fix them in the switch statement below, too!
```
Comment (markuspf) [inline arangod/Aql/OptimizerRules.cpp]: The modification PR has been merged.

File: arangod/Aql/ExecutionBlockImpl.cpp (hunk 1 above)
Comment (markuspf) [reply arangod/Aql/ExecutionBlockImpl.cpp]: Oh this is because you had to rename the parallel unsorted gather executor and we had a conversation about the unfortunate naming before, right?

Comment (goedderz) [reply arangod/Aql/ExecutionBlockImpl.cpp]: Yes, right.

File: arangod/Aql/OptimizerRules.cpp (hunk 3 above)
Comment (goedderz) [reply arangod/Aql/OptimizerRules.cpp]: Very good! Thanks.

**Comment by goedderz** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/7298/

**Comment by goedderz** [comment]
Enabling modification nodes leads to test failures. I'm moving this into another PR.

**Comment by goedderz** [comment]
Re-run due to merge conflicts:
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/7308/
